### PR TITLE
Fix #528: make bigram generation stopword aware

### LINKED ISSUE #528: Remove stopwords after creating bigrams to prevent incorrect bigrams
#### Description
I am dealing with a corpus containing the phrase "thank you very much" frequently.  Currently this results in a frequent bigram "thank much".  This results from generating the bigrams without a concept of removed stopwords.

#### Steps/Code to Reproduce
Create a word cloud containing frequent phrases having internal stopwords.

#### Expected Results
I expected one of two things.  Either that 1) "thank you", "very much" would be frequent bigrams (but not "you very", which consists entirely of bigrams.)  Or that 2) "thank" and "much" would show up as frequent unigrams.  I think the first option is the more elegant solution.

#### Actual Results
"thank much" was a frequent bigram.

#### Versions
```
>>> import wordcloud; print("wordcoud", wordcloud.__version__)
wordcoud 1.6.0
```

### PR DESCRIPTION
**Pull request by carlgieringer**
I was a little confused on the logic of `test_wordcloud.py::test_small_canvas`, and this fix broke it.  I think my update to it is logical, but I'm not 100% sure what the logic of asserting a single `layout_` element was.

### CODE REVIEW

**Comment by carlgieringer** [comment]
Weird, the tests pass for me locally.  Why do you think that this `test_small_canvas` would fail remotely?  Here is my system info:

```
$ python
Python 3.7.6 (default, Feb 22 2020, 21:07:37)
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform; print(platform.platform())
Darwin-19.3.0-x86_64-i386-64bit
>>> import sys; print("Python", sys.version)
Python 3.7.6 (default, Feb 22 2020, 21:07:37)
[Clang 11.0.0 (clang-1100.0.33.17)]
>>> import numpy; print("NumPy", numpy.__version__)
NumPy 1.18.1
>>> import matplotlib; print("matplotlib", matplotlib.__version__)
matplotlib 3.1.3
>>> import wordcloud; print("wordcoud", wordcloud.__version__)
wordcoud 1.6.0.post28+gf26dd63
```

**Comment by amueller** [comment]
Hey. Thank you for the contribution!
There was a recent fix in the small canvas test here: #525.
Can you make sure you have the current version of master? Sorry I'm a bit swamped but I'll try to review this soon.

File: wordcloud/tokenization.py [hunk 1]
```diff
@@ -51,17 +53,20 @@ def unigrams_and_bigrams(words, normalize_plurals=True):
...
-        # collocation detection (30 is arbitrary):
-        word1 = standard_form[bigram[0].lower()]
-        word2 = standard_form[bigram[1].lower()]
+        # either bigram word might be missing if it's a stopword.
+        word1 = standard_form.get(bigram[0].lower())
+        word2 = standard_form.get(bigram[1].lower())
 
-        if score(count, counts[word1], counts[word2], n_words) > 30:
+        # collocation detection (30 is arbitrary):
+        if not word1 or not word2 or score(count, counts[word1], counts[word2], n_words) > 30:
```
Comment (amueller) [inline wordcloud/tokenization.py]: Sorry, I don't understand this and/or might be missing something obvious. This always evaluated to True if one of them is a stopword, right? But it might still not be a collocation, right?

Comment (carlgieringer) [reply wordcloud/tokenization.py]: You're correct.  I read up about collocations and have updated the branch.

**Comment by carlgieringer** [comment]
Yes, my branch is up-to-date with your master, including #525.

**Comment by carlgieringer** [comment]
Weird.  Error looks unrelated to my changes:

```
BUILD FAILED (OS X 10.13.6 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/python-build.20200301213032.36590

Results logged to /var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/python-build.20200301213032.36590.log

Last 10 log lines:

(cd /Users/travis/.pyenv/versions/3.4.8/share/man/man1; ln -s python3.4.1 python3.1)

if test "xupgrade" != "xno"  ; then \

		case upgrade in \

			upgrade) ensurepip="--upgrade" ;; \

			install|*) ensurepip="" ;; \

		esac; \

		 ./python.exe -E -m ensurepip \

			$ensurepip --root=/ ; \

	fi

Ignoring ensurepip failure: pip 9.0.1 requires SSL/TLS

Traceback (most recent call last):

  File "../addons/travis/install_pyenv.py", line 132, in <module>

    install(os.environ['PYTHON_VERSION'])

  File "../addons/travis/install_pyenv.py", line 121, in install

    check_output(cmd, shell=True)

  File "/usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 223, in check_output

    raise CalledProcessError(retcode, cmd, output=output)

subprocess.CalledProcessError: Command 'eval "$( pyenv init - )"

pyenv install 3.4.8' returned non-zero exit status 1

A command failed while executing BEFORE_INSTALL step.

  Return code:

    1

  Command:

    python ../addons/travis/install_pyenv.py

  Working directory:

    /Users/travis/build/amueller/word_cloud
```

File: test/test_wordcloud.py [hunk 2]
```diff
@@ -41,19 +41,42 @@
...
+thanks
+"""
+
+SMALL_CANVAS = """
+better late than never someone will say
+"""
+
 
 def test_collocations():
-    wc = WordCloud(collocations=False, stopwords=[])
```
Comment (amueller) [inline test/test_wordcloud.py]: Is this change required?

**Comment by amueller** [comment]
Thanks for the update. I think the travis failure might be unrelated, but I have to double check. The diff seems like it contains unrelated content, though?

**Comment by amueller** [comment]
yeah the travis failure is unrelated. I should really update that....

**Comment by jcfr** [comment]
> yeah the travis failure is unrelated. I should really update that....

Issue has been fixed and this topic has been rebased.

File: test/test_wordcloud.py (hunk 2 above)
Comment (carlgieringer) [reply test/test_wordcloud.py:46]: Stopwords [is documented to be a set](https://github.com/amueller/word_cloud/blob/master/wordcloud/wordcloud.py#L204).  In Python `[]` is an empty array.  So I was updating it to use the correct type.

File: test/test_wordcloud.py [hunk 3]
```diff
@@ -352,8 +375,8 @@ def test_recolor_too_small_set_default():
 def test_small_canvas():
     # check font size fallback works on small canvas
     wc = WordCloud(max_words=50, width=21, height=21)
-    wc.generate(THIS)
-    assert len(wc.layout_) == 1
+    wc.generate(SMALL_CANVAS)
```
Comment (amueller) [inline test/test_wordcloud.py:378]: Is the SMALL_CANVAS change related or from a rebase?

Comment (carlgieringer) [reply test/test_wordcloud.py:378]: I'm not familiar with the rebase.  I did make changes to `SMALL_CANVAS` because this test didn't pass, even though my changes seem irrelevant to it.  We discussed this issue [here](https://github.com/amueller/word_cloud/pull/529#issuecomment-592735274)

Comment (jcfr) [reply test/test_wordcloud.py:378]: > from a rebase?

@amueller You can see the difference between the before and after rebase clicking on "force-pushed" link:

image

I this case, the differences are visible using this link: https://github.com/amueller/word_cloud/compare/bacd96e25e083b57378a4363746286dda32803a8..3983ef8056fb0254d2a6365811a09d3efe75fc43

Comment (amueller) [reply test/test_wordcloud.py:378]: The error should have been fixed in #525

**Comment by amueller** [comment]
also see issue #528.
